File: c/mergesort.c
#include <stdio.h>
#include <stdlib.h>

void merge(int [], int, int, int);
void sort(int[], int, int);
 
int main(int argc, char * argv[]) {
    int i, size;
    size = atoi(argv[1]);
    int A[size];
    for (i=0; i<size; i++) {
        scanf("%d", &A[i]);
    }
    sort(A, 0, size-1);
    for (i=0; i<size; i++) {
        printf("%d ", A[i]);
    }
    printf("\n");
    return 0;
}

void merge(int A[], int start, int mid, int end) {
    int sizeL = mid - start + 1;
    int sizeM = end - mid;
    int L[sizeL], M[sizeM];
    int i, j, k;
    /* Split the array A[] into L[] and M[] */
    for (i=0; i<sizeL; i++) {
        L[i] = A[start+i];
    }
    for (j=0; j<sizeM; j++) {
        M[j] = A[mid+j+1];
    }
    i = 0;
    j = 0;
    k = start;
    /* Sort the sub-arrays while merging back to A[] */
    while ((i < sizeL) && (j < sizeM)) {
        if (L[i] < M[j]) {
            A[k] = L[i];
            i++;
        }
        else {
            A[k] = M[j];
            j++;
        }
        k++;
    }
    /* If M[] exhausted, put remaining L[] back */
    while (i < sizeL) {
        A[k] = L[i];
        i++;
        k++;
    }
    /* If L[] exhausted, put remaining M[] back */
    while (j < sizeM) {
        A[k] = M[j];
        j++;
        k++;
    }
}

void sort(int A[], int start, int end) {
    /* Base case is a single element array i.e already sorted */
    if (start < end) {
        int mid = (start + end)/2;
        /* Divide and conquer ;) */
        sort(A, start, mid);
        sort(A, mid+1, end);
        merge(A, start, mid, end);
    }
}

File: python/merge-sort.py
#!/usr/bin/env python3

import sys

def main():
    arr = []
    fname = sys.argv[1]
    with open(fname, 'r') as f:
        for line in f:
            arr.append(int(line.rstrip('\r\n')))
    merge_sort(arr, start=0, end=len(arr)-1)
    print('Sorted list is: ', arr)
    return

def merge(arr, start, mid, end):
    # Slice the array from mid
    left = arr[start:mid+1]
    right = arr[mid+1:end+1]
    i = 0
    j = 0
    k = start
    # Sort and merge back the slices, until one exhausts
    while ((i < len(left)) and (j < len(right))):
        if (left[i] <= right[j]):
            arr[k] = left[i]
            i += 1
            k += 1
        else:
            arr[k] = right[j]
            j += 1
            k += 1
    # If right exhausted, merge back remaining left
    while (i < len(left)):
        arr[k] = left[i]
        i += 1
        k += 1
    # If left exhausted, merge back remaining right
    while (j < len(right)):
        arr[k] = right[j]
        j += 1
        k += 1
    return

def merge_sort(arr, start, end):
    if (start < end):
        mid = int((start + end) / 2)
        merge_sort(arr, start, mid)
        merge_sort(arr, mid+1, end)
        merge(arr, start, mid, end)
    return

if (__name__ == '__main__'):
    main()
